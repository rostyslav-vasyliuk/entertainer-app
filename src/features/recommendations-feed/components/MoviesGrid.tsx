import React from 'react';
import { View, Text, TouchableOpacity, Animated, Easing, ActivityIndicator } from 'react-native';
import { Image } from 'react-native-elements';
import { screenWidth } from '../../../constants/screen-contants';
import { TEXT_COLOR_SECONDARY, LOADER_COLOR } from '../../../constants/color-constants';

const GRID_AMOUNT = 6;

const MoviesGrid = ({ data, navigation, label }) => {

  let gridData = data.slice(0, GRID_AMOUNT);
  const navigate = (current_id) => {
    navigation.push('MovieDetails', {
      movie_id: current_id
    })
  }

  const getSmallGridItem = (element) => {
    const animatedValue = new Animated.Value(1);
    const onPressInAnimation = () => {
      Animated.timing(animatedValue, {
        toValue: 0.96, duration: 200, easing: Easing.ease
      }).start();
    }

    const onPressOutAnimation = () => {
      Animated.timing(animatedValue, {
        toValue: 1, duration: 200
      }).start();
    }

    return (
      <TouchableOpacity
        activeOpacity={0.9}
        onPressIn={() => onPressInAnimation()}
        onPressOut={() => onPressOutAnimation()}
        onPress={() => navigate(element.id)}
      >
        <Animated.View style={{ transform: [{ scale: animatedValue }] }}>
          <View style={{ width: (screenWidth / 2) - 16, borderRadius: 6, margin: 3, height: 220 }}>
            <Image
              source={{ uri: 'https://image.tmdb.org/t/p/w500/' + element.poster_path }}
              style={{ width: (screenWidth / 2) - 16, borderRadius: 6, height: 220 }}
              PlaceholderContent={<ActivityIndicator color={LOADER_COLOR} />}
              borderRadius={6}
            />
          </View>
        </Animated.View>
      </TouchableOpacity>
    )
  }

  const getBigGridItem = (element) => {
    const animatedValue = new Animated.Value(1);
    const onPressInAnimation = () => {
      Animated.timing(animatedValue, {
        toValue: 0.96, duration: 200, easing: Easing.ease
      }).start();
    }

    const onPressOutAnimation = () => {
      Animated.timing(animatedValue, {
        toValue: 1, duration: 200
      }).start();
    }

    return (
      <TouchableOpacity
        activeOpacity={0.9}
        onPressIn={() => onPressInAnimation()}
        onPressOut={() => onPressOutAnimation()}
        onPress={() => navigate(element.id)}
      >
        <Animated.View style={{ transform: [{ scale: animatedValue }] }}>
          <View style={{ width: (screenWidth / 2) - 16, borderRadius: 6, margin: 3, height: 300 }}>
            <Image
              source={{ uri: 'https://image.tmdb.org/t/p/w500/' + element.poster_path }}
              style={{ width: (screenWidth / 2) - 16, borderRadius: 6, height: 300, justifyContent: 'flex-end' }}
              PlaceholderContent={<ActivityIndicator color={LOADER_COLOR} />}
              borderRadius={6}
            />
          </View>
        </Animated.View>
      </TouchableOpacity>
    )
  }


  return (
    <View>
      <Text style={{ color: TEXT_COLOR_SECONDARY, paddingBottom: 6, paddingTop: 10, paddingLeft: 14, textTransform: 'uppercase', fontSize: 16, fontWeight: '600', letterSpacing: 1 }}>
        {label}
      </Text>

      <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'center', paddingBottom: 10 }}>
        <View>
          {gridData.map((element, index) => {
            if (GRID_AMOUNT / 2 > index) {
              if (index === 0) {
                return getSmallGridItem(element);
              }

              return getBigGridItem(element)
            }
          })}
        </View>

        <View>
          {gridData.map((element, index) => {
            if ((GRID_AMOUNT / 2) <= index) {
              if (index === GRID_AMOUNT - 1) {
                return getSmallGridItem(element);
              }

              return getBigGridItem(element)
            }
          })}
        </View>
      </View>
    </View>
  )
}

export default MoviesGrid;
