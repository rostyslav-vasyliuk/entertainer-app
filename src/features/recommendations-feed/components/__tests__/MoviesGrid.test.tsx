import React from 'react';
import MoviesGrid from '../MoviesGrid';
import renderer from 'react-test-renderer';

describe('<AppIntroOverlay />', () => {
  it('match snapshot', () => {
    const tree = renderer.create(<MoviesGrid navigation={{}} data={[]} label={''} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has 1 child', () => {
    const tree = renderer.create(<MoviesGrid navigation={{ navigate: () => { } }} data={[]} label={''} />).toJSON();
    expect(tree.children.length).toBe(2);
  });

  it('has 1 child', () => {
    const tree = renderer.create(<MoviesGrid navigation={{ navigate: () => { } }} data={[]} label={''} />).toJSON();
    expect(tree.children.length).toBeGreaterThanOrEqual(2);
  });
});
