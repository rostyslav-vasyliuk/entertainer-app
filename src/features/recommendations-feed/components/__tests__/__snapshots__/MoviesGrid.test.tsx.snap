// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppIntroOverlay /> match snapshot 1`] = `
<View>
  <Text
    style={
      Object {
        "color": "#c0c1c2",
        "fontSize": 16,
        "fontWeight": "600",
        "letterSpacing": 1,
        "paddingBottom": 6,
        "paddingLeft": 14,
        "paddingTop": 10,
        "textTransform": "uppercase",
      }
    }
  >
    
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 10,
      }
    }
  >
    <View />
    <View />
  </View>
</View>
`;
